File: src/com/lushprojects/circuitjs1/client/UndoManager.java
/*
    Copyright (C) Paul Falstad and Iain Sharp

    This file is part of CircuitJS1.

    CircuitJS1 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 2 of the License, or
    (at your option) any later version.

    CircuitJS1 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CircuitJS1.  If not, see <http://www.gnu.org/licenses/>.
*/

package com.lushprojects.circuitjs1.client;

import java.util.Vector;
import com.google.gwt.storage.client.Storage;

public class UndoManager {

    CirSim sim;
    Vector<UndoItem> undoStack, redoStack;

    UndoManager(CirSim sim) {
	this.sim = sim;
	undoStack = new Vector<UndoItem>();
	redoStack = new Vector<UndoItem>();
    }

    class UndoItem {
	public String dump;
	public double scale, transform4, transform5;
	UndoItem(String d) {
	    dump = d;
	    scale = sim.transform[0];
	    transform4 = sim.transform[4];
	    transform5 = sim.transform[5];
	}
    }

    void pushUndo() {
    	redoStack.removeAllElements();
    	String s = sim.dumpCircuit();
    	if (undoStack.size() > 0 &&
    			s.compareTo(undoStack.lastElement().dump) == 0)
    	    return;
    	undoStack.add(new UndoItem(s));
    	enableUndoRedo();
    	sim.savedFlag = false;
    }

    void doUndo() {
    	if (undoStack.size() == 0)
    		return;
    	redoStack.add(new UndoItem(sim.dumpCircuit()));
    	UndoItem ui = undoStack.remove(undoStack.size()-1);
    	loadUndoItem(ui);
    	enableUndoRedo();
    }

    void doRedo() {
    	if (redoStack.size() == 0)
    		return;
    	undoStack.add(new UndoItem(sim.dumpCircuit()));
    	UndoItem ui = redoStack.remove(redoStack.size()-1);
    	loadUndoItem(ui);
    	enableUndoRedo();
    }

    void loadUndoItem(UndoItem ui) {
	sim.loader.readCircuit(ui.dump, CircuitLoader.RC_NO_CENTER);
	sim.transform[0] = sim.transform[3] = ui.scale;
	sim.transform[4] = ui.transform4;
	sim.transform[5] = ui.transform5;
    }

    void doRecover() {
	pushUndo();
	sim.loader.readCircuit(sim.recovery);
	sim.allowSave(false);
	sim.menus.recoverItem.setEnabled(false);
    }

    void enableUndoRedo() {
    	sim.menus.redoItem.setEnabled(redoStack.size() > 0);
    	sim.menus.undoItem.setEnabled(undoStack.size() > 0);
    }

    void writeRecoveryToStorage() {
	sim.console("write recovery");
    	Storage stor = Storage.getLocalStorageIfSupported();
    	if (stor == null)
    		return;
    	String s = sim.dumpCircuit();
    	stor.setItem("circuitRecovery", s);
    }

    void readRecovery() {
	Storage stor = Storage.getLocalStorageIfSupported();
	if (stor == null)
		return;
	sim.recovery = stor.getItem("circuitRecovery");
    }
}

File: UndoManager.ts
/*
    Copyright (C) Paul Falstad and Iain Sharp

    This file is part of CircuitJS1.

    CircuitJS1 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 2 of the License, or
    (at your option) any later version.

    CircuitJS1 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CircuitJS1.  If not, see <http://www.gnu.org/licenses/>.
*/

import { CirSim } from "./CirSim";
import { CircuitLoader } from "./CircuitLoader";

class UndoItem {
    dump: string;
    scale: number;
    transform4: number;
    transform5: number;

    constructor(sim: CirSim, dump: string) {
        this.dump = dump;
        this.scale = sim.transform[0];
        this.transform4 = sim.transform[4];
        this.transform5 = sim.transform[5];
    }
}

export class UndoManager {
    sim: CirSim;
    undoStack: UndoItem[];
    redoStack: UndoItem[];

    constructor(sim: CirSim) {
        this.sim = sim;
        this.undoStack = [];
        this.redoStack = [];
    }

    pushUndo(): void {
        this.redoStack.length = 0;
        const s = this.sim.dumpCircuit();
        if (this.undoStack.length > 0 &&
                s === this.undoStack[this.undoStack.length - 1].dump)
            return;
        this.undoStack.push(new UndoItem(this.sim, s));
        this.enableUndoRedo();
        this.sim.savedFlag = false;
        this.sim.unsavedChanges = true;
    }

    doUndo(): void {
        if (this.undoStack.length === 0)
            return;
        this.redoStack.push(new UndoItem(this.sim, this.sim.dumpCircuit()));
        const ui = this.undoStack.pop()!;
        this.loadUndoItem(ui);
        this.enableUndoRedo();
    }

    doRedo(): void {
        if (this.redoStack.length === 0)
            return;
        this.undoStack.push(new UndoItem(this.sim, this.sim.dumpCircuit()));
        const ui = this.redoStack.pop()!;
        this.loadUndoItem(ui);
        this.enableUndoRedo();
    }

    loadUndoItem(ui: UndoItem): void {
        this.sim.loader.readCircuit(ui.dump, CircuitLoader.RC_NO_CENTER);
        this.sim.transform[0] = this.sim.transform[3] = ui.scale;
        this.sim.transform[4] = ui.transform4;
        this.sim.transform[5] = ui.transform5;
    }

    doRecover(): void {
        this.pushUndo();
        this.sim.loader.readCircuit(this.sim.recovery!);
        this.sim.allowSave(false);
        this.sim.menus.recoverItem.setEnabled(false);
        this.sim.unsavedChanges = false;
    }

    enableUndoRedo(): void {
        this.sim.menus.redoItem.setEnabled(this.redoStack.length > 0);
        this.sim.menus.undoItem.setEnabled(this.undoStack.length > 0);
    }

    writeRecoveryToStorage(): void {
        CirSim.console("write recovery");
        const stor = window.localStorage;
        if (stor == null)
            return;
        const s = this.sim.dumpCircuit();
        stor.setItem("circuitRecovery", s);
    }

    readRecovery(): void {
        const stor = window.localStorage;
        if (stor == null)
            return;
        this.sim.recovery = stor.getItem("circuitRecovery");
    }
}
